### ISSUE TITLE
How to export yolo as tensorflow model, that can be futher trained?

### ISSUE THREAD

**Issue by beenotung** [ultralytics/ultralytics#21147]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Is it possible to export yolo model, for further training in js?

What I know:
- training yolo model via yolo cli (python env)
- export yolo model as tfjs (for inference only in js env)

What I've tried:
- export yolo model as tfjs / tensorflow saved model (only can export as graph model, not layered model, which cannot be trained)
- convert onnx model as saved model, then save it as h5/keras model (it only contains weight, no architecture)

What I'm looking for:
- a way to convert yolo model into tensorflow.js (tfjs) layered model that can be trained
- or a way to train the onnx model in js env (if possible)
- or a way to train the .pt model in js env (if possible)

**Comment by UltralyticsAssistant** [comment]
👋 Hello @beenotung, thank you for reaching out and for your interest in Ultralytics 🚀! This is an automated response to help you get started—an Ultralytics engineer will also assist you soon.

We recommend reviewing the [Docs](https://docs.ultralytics.com/) for export and deployment workflows, as well as checking out many [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common export questions are covered there.

If this is a 🐛 Bug Report, please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) to help us debug it.

If this is a custom training ❓ Question (e.g., about further training models in TensorFlow.js), please include as much detail as possible about your dataset, code, logs, and any specific errors you've encountered. This will help us understand your use case and provide better guidance.

Join the Ultralytics community where it suits you best:
- For real-time chat, head to [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, visit [Discourse](https://community.ultralytics.com/)
- For sharing and exploring ideas, check out our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

Before proceeding, please upgrade to the latest `ultralytics` package and all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment with [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/) to ensure you have the newest features and fixes:

```bash
pip install -U ultralytics
```

## Environments

YOLO can be run in any of these up-to-date, verified environments (with all dependencies including [CUDA](https://developer.nvidia.com/cuda-zone)/[CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/) and [PyTorch](https://pytorch.org/) preinstalled):

- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are currently passing. CI tests verify correct operation of all YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

Thank you again for your detailed question and patience. An Ultralytics engineer will review and respond to your request as soon as possible!

**Comment by glenn-jocher** [comment]
@beenotung Unfortunately, exporting YOLO models to a trainable TensorFlow.js format is not currently supported. The [TF.js export functionality](https://docs.ultralytics.com/integrations/tfjs/) creates inference-optimized static graphs that cannot be further trained, which is why you're only getting graph models rather than layered models.

This limitation exists because YOLO's complex architecture, including anchor generation, loss functions, and post-processing steps, doesn't translate well to TF.js's training environment. The export process fundamentally converts the dynamic PyTorch model into an optimized static representation for inference.

For training in web environments, you'd need to either keep the training pipeline in Python and only deploy inference models to JavaScript, or consider using TensorFlow.js to build and train simpler detection models from scratch rather than trying to convert YOLO architectures.

**Comment by beenotung** [comment]
@glenn-jocher thanks for the clarification, I appreciate the efficient and high accurate result archived from transfer learning yolo model. I'll explore ways to invoke python code that interact with yolo in interactive manner. This approach will require a python runtime along side the web environment, which is less portable but at least possible 💪

**Comment by glenn-jocher** [comment]
@beenotung You're welcome! That sounds like a solid approach - running Python alongside your web environment via subprocess calls or a REST API will give you the full power of YOLO's training capabilities while still integrating with your JavaScript application. Good luck with your implementation! 👍
